### PR TITLE
asm-lsp: init at v0.4.2

### PR DESCRIPTION
**Pull request by NotAShelf**
## Description of changes

Adds asm-lsp

https://github.com/bergercookie/asm-lsp

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by eclairevoyant** [comment]
Getting a FOD hash mismatch:
```
error: hash mismatch in fixed-output derivation '/nix/store/1afywf0h7116v2m619jw5f2lw4y4h518-source.drv':
         specified: sha256-jVWZ0eOivDnpyo6BYef40tnl8roRbfXhA4eZ/Sx0VbU=
            got:    sha256-EGgYOU6y23ULjnMGNjYhgF0JMPgvRuQ4UOWqwJxhBpU=
error: 1 dependencies of derivation '/nix/store/33bx9iz77xgprj4qp5gjsj3f3484ks4y-asm-lsp-0.4.2-vendor.tar.gz.drv' failed to build
error: 2 dependencies of derivation '/nix/store/c999jv5r7a4xndyxhj5sn4kq76hbdm6h-asm-lsp-0.4.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/40xmrv47rghvavlxz394wwk38977j6i3-review-shell.drv' failed to build
```

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    rev = "v${version}";
+    hash = "sha256-jVWZ0eOivDnpyo6BYef40tnl8roRbfXhA4eZ/Sx0VbU=";
+  };
+
+  nativeBuildInputs = [
+    pkg-config
+  ];
+
+  buildInputs =
+    [
```
Comment (eclairevoyant) [inline pkgs/by-name/as/asm-lsp/package.nix:28]: I'd also move this up a line, for consistency

**Comment by NotAShelf** [comment]
> Getting a FOD hash mismatch:
> 
> ```
> error: hash mismatch in fixed-output derivation '/nix/store/1afywf0h7116v2m619jw5f2lw4y4h518-source.drv':
>          specified: sha256-jVWZ0eOivDnpyo6BYef40tnl8roRbfXhA4eZ/Sx0VbU=
>             got:    sha256-EGgYOU6y23ULjnMGNjYhgF0JMPgvRuQ4UOWqwJxhBpU=
> error: 1 dependencies of derivation '/nix/store/33bx9iz77xgprj4qp5gjsj3f3484ks4y-asm-lsp-0.4.2-vendor.tar.gz.drv' failed to build
> error: 2 dependencies of derivation '/nix/store/c999jv5r7a4xndyxhj5sn4kq76hbdm6h-asm-lsp-0.4.2.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/40xmrv47rghvavlxz394wwk38977j6i3-review-shell.drv' failed to build
> ```

ah I forgot to update the hash for 0.4.2 probably, lemme check

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  version = "0.4.2";
+in
+rustPlatform.buildRustPackage {
+  inherit pname version;
+
+  src = fetchFromGitHub {
+    owner = "bergercookie";
+    repo = "asm-lsp";
+    rev = "v${version}";
+    hash = "sha256-jVWZ0eOivDnpyo6BYef40tnl8roRbfXhA4eZ/Sx0VbU=";
```
Comment (eclairevoyant) [inline pkgs/by-name/as/asm-lsp/package.nix:20]:
```suggestion
    hash = "sha256-EGgYOU6y23ULjnMGNjYhgF0JMPgvRuQ4UOWqwJxhBpU=";
```

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  buildInputs =
+    [
+      openssl
+    ]
+    ++ lib.optionals stdenv.isDarwin [
+      darwin.apple_sdk.frameworks.Security
+    ];
+
+  cargoHash = "sha256-szchvIraeZEHDQSObtXLMZY0gFSg2vShorSSuSLL6MU=";
```
Comment (eclairevoyant) [inline pkgs/by-name/as/asm-lsp/package.nix:35]:
```suggestion
  cargoHash = "sha256-x8Cj39Wki+pdoNIO8QPGK29KFJrHtDMoZJIXFEldno0=";
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/by-name/as/asm-lsp/package.nix (hunk 1 above)
Comment (NotAShelf) [reply pkgs/by-name/as/asm-lsp/package.nix:28]: by "up a line do you mean something like this

```nix
 nativeBuildInputs = [
    pkg-config
  ];
  buildInputs =
    [
      openssl
    ]
    ++ lib.optionals stdenv.isDarwin [
      darwin.apple_sdk.frameworks.Security
    ];
```

**Review by eclairevoyant — APPROVED** [review]
Result of `nixpkgs-review pr 262602` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>asm-lsp</li>
  </ul>
</details>

Didn't see any issues running the binary either, haven't tested LSP functionality per se

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  nativeBuildInputs = [
+    pkg-config
+  ];
+
+  buildInputs = [
+    openssl
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.Security
```
Comment (fufexan) [inline pkgs/by-name/as/asm-lsp/package.nix:30]: You should probably only import `Security` and have it inherited in the `callPackage` call.

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  ];
+
+  cargoHash = "sha256-x8Cj39Wki+pdoNIO8QPGK29KFJrHtDMoZJIXFEldno0=";
+
+  meta = {
+    description = "Language server for NASM/GAS/GO Assembly";
+    homepage = "https://github.com/bergercookie/asm-lsp";
+    license = lib.licenses.bsd2;
+    maintainers = with lib.maintainers; [ NotAShelf ];
+    mainProgram = "asm-lsp";
```
Comment (fufexan) [inline pkgs/by-name/as/asm-lsp/package.nix:40]: `pname` can be used here.

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  buildInputs = [
+    openssl
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.Security
+  ];
+
+  cargoHash = "sha256-x8Cj39Wki+pdoNIO8QPGK29KFJrHtDMoZJIXFEldno0=";
+
+  meta = {
```
Comment (fufexan) [inline pkgs/by-name/as/asm-lsp/package.nix:35]: Missing `platforms`.

**Review by fufexan — CHANGES_REQUESTED** [review]

File: pkgs/by-name/as/asm-lsp/package.nix (hunk 4 above)
Comment (NotAShelf) [reply pkgs/by-name/as/asm-lsp/package.nix:30]: this is something nix-init does, I assume it has a purpose?

but looks like the program does not work on darwin anyway, so probably can be removed as a whole

File: pkgs/by-name/as/asm-lsp/package.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
+{ lib
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, openssl
+, stdenv
```
Comment (fufexan) [inline pkgs/by-name/as/asm-lsp/package.nix:6]: No longer needed unless I'm blind.

**Review by fufexan — APPROVED** [review]

**Review by Henry-Hiles — APPROVED** [review]
LGTM

**Comment by Freed-Wu** [comment]
Comes from <https://github.com/bergercookie/asm-lsp/issues/43>. How about update it to latest 0.6.0?

**Comment by NotAShelf** [comment]
Should be as simple as updating the version string and the hashes. I'll try to push an update shortly.

*P.S. in the future please open a package update request, I almost cleared my notifications since this was a closed issue*

**Comment by NotAShelf** [comment]
@Freed-Wu 

https://github.com/NixOS/nixpkgs/pull/275471

**Pull request merged**